File: backend/src/services/workLocationService.js
import { readJson, writeJson } from "../storage/jsonStore.js";
import { getSettings } from "./settingsService.js";
import { LOCATIONS } from "../constants/locations.js";

const WORK_LOCATIONS_FILE = "workLocations.json";

function getAllRecords() {
    return readJson(WORK_LOCATIONS_FILE, []);
}

function findRecord(userId, date) {
    const all = getAllRecords();
    return all.find((r) => r.userId === userId && r.date === date) || null;
}

function isInCompanyWfhPeriod(date) {
    const settings = getSettings();
    return settings.companyWfhPeriods.some(
        (p) => date >= p.startDate && date <= p.endDate
    );
}

/**
 * Resolution order:
 * 1. Individual record exists → use it
 * 2. Company WFH period active → WFH
 * 3. No data → OFFICE (default)
 */
export function getEffectiveLocation(userId, date) {
    const record = findRecord(userId, date);
    if (record) return record.location;

    if (isInCompanyWfhPeriod(date)) return LOCATIONS.WFH;

    return LOCATIONS.OFFICE;
}

export function setLocation(userId, date, location, updatedBy) {
    const all = getAllRecords();
    const existing = all.find(
        (r) => r.userId === userId && r.date === date
    );

    if (existing) {
        existing.location = location;
        existing.updatedBy = updatedBy;
        existing.updatedAt = new Date().toISOString();
    } else {
        all.push({
            userId,
            date,
            location,
            updatedBy,
            updatedAt: new Date().toISOString(),
        });
    }

    writeJson(WORK_LOCATIONS_FILE, all);
}

export function getLocationRecord(userId, date) {
    return findRecord(userId, date);
}

/**
 * Count WFH days per user for a calendar month.
 *
 * @param {Array} userList  - array of user objects { id, name, teamId }
 * @param {string} month    - "YYYY-MM"
 * @param {number} allowance - monthlyWfhAllowance from settings
 * @returns {Array} [{ userId, name, teamId, wfhDays, overLimit, extraDays? }]
 */
export function getMonthlyWfhUsage(userList, month, allowance) {
    const all = getAllRecords();

    return userList.map((user) => {
        const wfhDays = all.filter(
            (r) => r.userId === user.id && r.location === LOCATIONS.WFH && r.date.startsWith(month)
        ).length;

        const overLimit = wfhDays > allowance;
        const entry = { userId: user.id, name: user.name, teamId: user.teamId, wfhDays, overLimit };
        if (overLimit) entry.extraDays = wfhDays - allowance;
        return entry;
    });
}

File: mhp-python/app/services/work_location_service.py
"""
Work location service.
PK=DATE#<date>, SK=LOC#<userId>

Location resolution order (same as Node.js):
  1. Individual record exists → use it
  2. Company WFH period active → WFH
  3. Default → OFFICE
"""

from datetime import date as DateType, datetime, timezone
from app.database import get_item, put_item, query_by_pk, date_pk, location_sk
from app.constants import Locations
from app.services.settings_service import get_settings


def get_effective_location(user_id: str, date_str: str) -> str:
    item = get_item(date_pk(date_str), location_sk(user_id))
    if item:
        return item["location"]

    settings = get_settings()
    for period in settings.get("companyWfhPeriods", []):
        if period["startDate"] <= date_str <= period["endDate"]:
            return Locations.WFH

    return Locations.OFFICE


def set_location(user_id: str, date_str: str, location: str, updated_by: str):
    item = {
        "PK": date_pk(date_str),
        "SK": location_sk(user_id),
        "userId": user_id,
        "date": date_str,
        "location": location,
        "updatedBy": updated_by,
        "updatedAt": datetime.now(timezone.utc).isoformat(),
    }
    put_item(item)


def get_location_record(user_id: str, date_str: str) -> dict | None:
    item = get_item(date_pk(date_str), location_sk(user_id))
    if not item:
        return None
    item.pop("PK", None)
    item.pop("SK", None)
    return item


def get_all_locations_for_date(date_str: str) -> list[dict]:
    """All location records for a date — used by headcount service."""
    items = query_by_pk(date_pk(date_str), "LOC#")
    for item in items:
        item.pop("PK", None)
        item.pop("SK", None)
    return items


def get_monthly_wfh_usage(user_list: list[dict], month: str, allowance: int) -> list[dict]:
    """
    Count WFH days per user for a calendar month.
    month = "YYYY-MM"
    Note: we scan each user's dates for the month. This works fine at 100 users.
    """
    from app.database import scan_with_filter
    from boto3.dynamodb.conditions import Attr

    # Fetch all LOC records for the month in one scan
    items = scan_with_filter(
        Attr("SK").begins_with("LOC#") & Attr("date").begins_with(month)
    )

    # Build a map: userId → wfh day count
    wfh_counts: dict[str, int] = {}
    for item in items:
        if item.get("location") == Locations.WFH:
            uid = item["userId"]
            wfh_counts[uid] = wfh_counts.get(uid, 0) + 1

    result = []
    for user in user_list:
        wfh_days = wfh_counts.get(user["id"], 0)
        over_limit = wfh_days > allowance
        entry = {
            "userId": user["id"],
            "name": user["name"],
            "teamId": user.get("teamId"),
            "wfhDays": wfh_days,
            "overLimit": over_limit,
        }
        if over_limit:
            entry["extraDays"] = wfh_days - allowance
        result.append(entry)
    return result
